concrete/src/Api/OAuth/Client/ClientFactory.php

```diff
@@ -0,0 +1,102 @@
+<?php
+namespace Concrete\Core\Api\OAuth\Client;
+
+use Concrete\Core\Application\Application;
+use Concrete\Core\Entity\OAuth\Client;
+use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\Id\UuidGenerator;
+
+class ClientFactory
+{
+
+    /**
+     * @var \Concrete\Core\Application\Application
+     */
+    private $app;
+
+    /**
+     * @var \Doctrine\ORM\EntityManager
+     */
+    private $entityManager;
+
+    /**
+     * @var \Doctrine\ORM\Id\UuidGenerator
+     */
+    private $generator;
+
+    public function __construct(Application $app, EntityManager $entityManager, UuidGenerator $generator)
+    {
+        $this->app = $app;
+        $this->entityManager = $entityManager;
+        $this->generator = $generator;
+    }
+
+    /**
+     * Create a new OAuth client object and provide it a proper UUID
+     *
+     * @param string $name The client name
+     * @param string $redirect The redirect url
+     * @param string[] $scopes The scopes this client is allowed to interact with
+     * @param string $key The client's api key
+     * @param string $secret This secret should be properly hashed with something like `password_hash`
+     * @return \Concrete\Core\Entity\OAuth\Client
+     */
+    public function createClient($name, $redirect, array $scopes, $key, $secret)
+    {
+        $client = $this->app->make(Client::class);
+        $client->setName($name);
+        $client->setRedirectUri($redirect);
+        // @TODO support scopes
+        //$client->setScopes($scopes);
+        $client->setClientKey($key);
+        $client->setClientSecret($secret);
+
+        // Apply a new UUID to the client
+        $id = $this->generator->generate($this->entityManager, $client);
+        $client->setIdentifier($id);
+
+        return $client;
+    }
+
+    /**
+     * Generate new credentials for use with a client
+     * Note: the secret provided by with these credentials is in plain text and should be hashed before storing to the
+     * database.
+     *
+     * @param int $keyLength
+     * @param int $secretLength
+     * @return \Concrete\Core\Api\OAuth\Client\Credentials
+     * @throws \InvalidArgumentException If an invalid size is passed
+     */
+    public function generateCredentials($keyLength = 64, $secretLength = 96)
+    {
+        // Make sure we have ints
+        $keyLength = (int) $keyLength;
+        $secretLength = (int) $secretLength;
+
+        // Make sure our keys are long enough
+        if ($keyLength < 16 || $secretLength < 16) {
```

Comment:
I wonder how this validation works. It compares with the value, not the length, right?

Reply:
No this method takes 2 integers, `$keyLength` and `$secretLength` and fails if either integer is less than 16. 
We've already tested our key generation method and we can assume it will give us good output if we give it good input, so we don't need to actually validate anything other than the input we receive.